### ISSUE TITLE
Create a Timestamp cell

### ISSUE THREAD

**Issue by braxtonmckee** [APrioriInvestments/nativepython#118]
This is for displaying timestamps. Ideally they just produce a nice format in the user's timezone (and make it clear what the timezone is on mouseover), with something to make it easy to see what's going on.

**Comment by dkrasner** [comment]
do you want this to display the current time in the user's timezone or to display any timestamp string coming from the server side? if it's the latter this is just a string display which happens to be a timestamp

independent of the above is there a format you like, ex `%Y-%m%d-%h:%s %Z`?

**Comment by braxtonmckee** [comment]
On the server side it will be a posix timestamp (seconds since the epoch,
utc). Javascript should be doing the conversion to an actual date.
Normally I just do what python's datetime library would do (e.g.
'2019-06-26 12:41:38.122146'), in the current machine timezone.  It would
be nice if there were some color coding so that the milliseconds were
lighter colored (sometimes hard to parse what the timestamp really is).
Mouseover should tell you how many seconds ago this was, and what timezone
is being displayed.

**Comment by dkrasner** [comment]
ok so client side current time is what to be displayed

This is at load time or this is something that is continually ticking?

`Mouseover should tell you how many seconds ago this was`  what is `this`?

**Comment by braxtonmckee** [comment]
Yes, client side converts the timestamp to a user string.  It's not ticking
(it just accepts a float). If you want it to tick you'd do a
`cells.Subscribed(lambda: cells.Timestamp(someTimestamp()))` which would
produce a new Timestamp cell each time it updates.

On the mouseover, the idea is that if I hover over the timestamp, i want a
tooltip or something to show me how long ago this was in a user-friendly
way (e.g. 10 seconds ago). Maybe that's no so great, but I thought it might
be useful since sometimes its hard to tell what's going on when you have a
whole grid of timestamps...

**Comment by dkrasner** [comment]
ok just to confirm:

server side sends over seconds since epoch (an integer); client converts that to local time for the user 

client side displays a "nice" string of that timestamp from year to milliseconds

on hover over user can see his current timezone 

on hover over use can see the difference between current time and the displayed time (have to think about how to do this without making things too complicated)

**Comment by braxtonmckee** [comment]
yes, except that server will send a float

**Comment by dkrasner** [comment]
ok check out the Timestamp cell examples in the playground (https://github.com/APrioriInvestments/nativepython/tree/daniel-cells-playground-1 branch) 

doesn't have hover over time diff (need to add some callbacks for that) but shows local tz on hover

**Comment by braxtonmckee** [comment]
That looks pretty good! I don't think we want the explicit div styling to
show up in the final API (cells should abstract the browserness away) but
that looks good.

**Comment by dkrasner** [comment]
i can remove div styling from these, but that might lead to issues preventing you from specifying colors, text, sizes etc

unless this is a cell that is going to show in a small number of layouts always  in the same way that is... 

if you have such a handful of examples (screenshots or sketches) I could cater the api

in the meantime i'll just remove it and we can go from there

**Comment by braxtonmckee** [comment]
So this is an interesting point.  I was thinking that what we should do is make an API that can specify things like text, color, and layout properties as general-purpose modifiers. For instance, one model could be to allow things like

```
from cells import Card, Text, styling

styling.background("green") * Card(
    styling.textSize(9) * Text("Some Text") + 
    styling.bold() * Text("Some bold text below.")
)
```

with the idea that a 'styling' verb can be applied as a modifier to everything below it. What I'm really hoping for is to have a bunch of verbs for styling that can apply to a broad class of cells, instead of applying methods on each of them, and also to prevent a bunch of style strings floating around in the code.  Implementation could simply carry them from cells to children, or do it all in css, I don't know which is better.

What do you think?

**Comment by dkrasner** [comment]
we could try to architect something like that, but it would take some thought

here are some issues to think about there:

* styling that is split between two frameworks (server and client) will create confusion, i.e. you will never be quite sure why some DOM element is taking on a particular style, and you would be forced to dig through two codebases to sort this out
* a related point is inline or in-code styling. The question is do you allow styling primarily in the code itself (like you propose), or do you keep css completely separate and referring to defined css classes (which then encapsulate the structure you expect within parents, children or siblings), sometimes allowing for "emergency" overrides such as dynamically defining a style change simply b/c it's much cleaner to do right there in the component itself. 
  * Every framework I have seen (React, Elm, even Seaside/Smalltalk) starts off with the former and **every** one then drops it in a subsequent major release. 
  * again it becomes havoc trying to track down culprits when the styling code is scattered all over the code base 
  * and it becomes more complicated to use the natural css functional structure (which is quite powerful) 
* when you say `modifier to everything below it` that will not work uniformly or in way you assume. Some style attribute like `background_color` might or might be inherited; for example, if you child is an `input` element - the input area will not inherit this background color using some default. Of course you could specify this in a css class but saying that an `input` tag child element of my parent css class must inherit background color ... so this could lead to the server side developer saying "huh why doesn't this propagate"  and potentially having no way to effectively force it so


in the meantime, I have removed `divStyle` from the Timestamp - should i merge [the pr](https://github.com/APrioriInvestments/nativepython/pull/132)

**Comment by braxtonmckee** [comment]
I totally get your point that the inline styling is a problem because then
you have no ability to figure out where some particular piece of styling is
coming from. But the python programmer has no access to css directly - the
way this API works you're not supposed to create new components, you're
supposed to combine existing pieces to build what you want.

So another approach to this would be to create a set of more explicit
modifiers that achieve specific effects but that don't have "styling" in
them. For instance, instead of writing `cells.Card("some
stuff").backgroundColor("light-grey")`, we could have `cells.inactive() *
cells.Card("some stuff")`, with the idea that something like
'cells.inactive` is a modifier that simply places a css style on the
resulting object in the browser, and we come up with a set of modifiers
that let us do what we need to do, but have the actual styling implemented
in css in a single place.  This would be better because 'cells.inactive()'
actually carries some semantic content (we want this widget to look
inactive), but defers making it look that way to the framework.

Does that seem like a viable approach?

**Comment by dkrasner** [comment]
i think this is a discussion for a different PR/place, but essentially the point is that if you adhere to an in-code style/structure you will forgo the functional framework that css has to offer. That's a choice. 

i think we should not have any css or css like definitions on the server/python side and all for only simple, declarative, cell api arguments such as `fit` `stretch` etc and have the client deal with all actual styles

**Comment by dkrasner** [comment]
close?

**Comment by braxtonmckee** [comment]
yes
